### ISSUE TITLE
Add option for LightningModule.log() to track by number of seen examples and not number of seen batches

### ISSUE THREAD

**Issue by jaymody** [PyTorchLightning/pytorch-lightning#5987]
Given:
```
num_epochs = 10
batch_size = 64
```
`LightningModule.log()` will log a metric to Tensorboard with step values of:

`1,2,3,4,...,639,640`

These values represent the number of batches the model has seen (`current_epoch*batch_idx`). It would be nice to instead have the option to track number of seen examples:

`64,128,192,256,...,40896,40960`

This can be computed via `current_epoch*batch_idx*batch_size` (same as before, but multiplied by the batch size).

### Motivation

When changing the batch size in training experiments, the step value no longer provides a one-to-one comparison. The next best thing is to use the "relative" feature in Tensorboard, which alters the x-axis to represent time, however this is not ideal and will break down when changing certain hyperparameters that affect training time, or when training on different compute resources.

### Pitch

Something along the lines of:
```
    def training_step(self, batch, batch_idx):
        ...
        self.log("mymetric", myvalue, step_by_nexamples=True)
        ...
```

For more flexibility, letting the user specify step would be nice (although I'm not sure how one would get the current epoch number to compute the number of examples):

```
    def training_step(self, batch, batch_idx):
        ...
        self.log("mymetric", myvalue, step=epoch_number*batch_idx*len(batch))
        ...
```

**Comment by carmocca** [comment]
Per-step is the smallest granularity available in the neural network training process (not just for lightning) as all computations are batched for efficiency. If your logged value is an aggregate over a batch, it doesn't make sense to log it per example.

Otherwise, if your logged value is per batch sample, you could always log it separately with a loop.

**Comment by jaymody** [comment]
The logging is still happening at every batch step, just as before (as you said, it is the smallest granularity avaliable). However the `global_step` value passed to Tensorboard can by anything you want. Right now in pytorch lightning, that value represents the number of batches the model has seen. If you were to multiply it by `batch_size`, it would represents the number of examples the model has seen. This is useful because it makes it significantly easier to compare training runs across different batch size. For example, consider the following two experiments (x-axis represents step):

<img width="1056" alt="Screen Shot 2021-02-15 at 2 29 46 PM" src="https://user-images.githubusercontent.com/26451316/107986588-61b79500-6f9a-11eb-94df-91d27211b014.png">

The grey model used a batch size of `512`, the blue used a batch size of `128`. Without knowing the batch sizes, it seems that the grey model trained faster and required fewer training examples to reach the same performance. However this is not the case. Both models saw the **exact same** amount of data, but since the step represents batch size, it is skewed to look that way. For this reason, when I write my training loops, I log tensorboard with `tb_writer.log("metric", val, global_step=batch_num*epoch*batch_size)` so that I can easily compare experiments without having to worry about how batch size is affecting the step. Another option is to use the "relative" view feature in Tensorboard which normalizes the x-axis for time:

<img width="1056" alt="Screen Shot 2021-02-15 at 2 30 01 PM" src="https://user-images.githubusercontent.com/26451316/107986782-c541c280-6f9a-11eb-845a-d5792e29a38e.png">

However, changes in hyperparameters, compute resources, implementation, etc ... will make training times differ, so this is not a reliable measure.

Hope this makes sense.

**Comment by carmocca** [comment]
> Without knowing the batch sizes, it seems that the grey model trained faster and required fewer training examples to reach the same performance

It might seem like that for an untrained eye who doesn't read axis labels 😛. But as long as the x-axis is annotated as steps, steps are just batches processed, they contain no information about time taken. Anybody assuming that one trained faster by looking at the graph is wrong. It is just a graph design matter.

Now, even if you logged per example, comparing plots might be problematic if the data does not fit well its underlying distribution (or variable length...) since training dataloaders are shuffled.

I do get what you are proposing but doesn't seem to me like it is something we should change in Lightning when it depends highly on the data/problem (my opinion).

Logically, you are free to log as you like in your code.
